Rover/AP_Scripting: scripting support for direct steering and throttle control

rmackay9 opened the pull request:
This PR add support for Lua scripts to directly control a Rover's steering and throttle.

It's pretty straight forward but:

- AP_Vehicle and Rover implement a new set_desired_steering_and_throttle() method.  The steering is in the (odd) range of -4500 to +4500 and throttle is in the range -100 to +100.  I'm happy to change these ranges to be -1 to +1 if others would prefer this.
- AP_Scripting's bindings make the new method visible to scripts
- AP_Scripting gets a new example script, rover-set-steering-and-throttle.lua

This has been tested in SITL and below is a screen shot of the script running successfully
![image](https://user-images.githubusercontent.com/1498098/84641498-0f6b3380-af36-11ea-9b08-e6f226d73597.png)

rmackay9 commented:
Rebased on master and tested and it still works.

rmackay9 commented:
@peterbarker, @khancyr, if you guys have a chance to review this that would be great.  No huge pressure though..

peterbarker commented:
Currently conflicting.

rmackay9 commented:
@peterbarker, I've resolved the merge conflict, it was caused by a recent merged PR re dynamic notch filters.

WickedShell commented:
Yeah, I'd strongly agree that lets do everything possible to avoid leaking the 4500/100 abstractions in here, and move straight to a float with a range to 1.0. That would let us remap the range in ArduPilot freely without breaking scripts. A key example of why this matters for example is that plane has a range on the throttle of 0-100, and you actually get observable effects from a change of 1, and it can result in jitter on RC input causing observable oscillations.

rmackay9 commented on Rover/mode.h:
```diff
@@ -416,6 +420,12 @@ class ModeGuided : public Mode
     float _desired_speed;       // desired speed used only in HeadingAndSpeed submode
 
+    // direct steering and throttle control
+    bool _have_strthr_target;       // true if we have a valid steering and throttle target
```
I've removed the "_target" part of the variable name, txs!

rmackay9 commented on libraries/AP_Vehicle/AP_Vehicle.h:
```diff
@@ -172,10 +172,12 @@ class AP_Vehicle : public AP_HAL::HAL::Callbacks {
     // get target location (for use by scripting)
     virtual bool get_target_location(Location& target_loc) { return false; }
 
+    // set desired steering (-4500 to +4500) and throttle (-100 to +100) (for use by scripting)
```
I've added a comment, "for use by scripting with Rover".

rmackay9 commented:
I've removed "desired" from everywhere and also change the scaling to be -1 to +1 for both steering and throttle.  Also rebased on master, thanks for the reviews!

rmackay9 commented on Rover/mode.h:
```diff
@@ -416,6 +420,12 @@ class ModeGuided : public Mode
+    // direct steering and throttle control
+    bool _have_strthr_target;       // true if we have a valid steering and throttle target
+    uint32_t _des_strthr_time_ms;   // system time last call to set_desired_steering_and_throttle was made (used for timeout)
+    float _des_strthr_steering;     // desired steering in the range -100 to +100
```
I've removed the _des_ part, thanks!

rmackay9 commented:
Merged, thanks for the reviews!
